File: adaptive_repair/data/python_programs/topological_ordering.py
import collections

def topological_ordering(nodes):
    """
    Topological Sort using Kahn's algorithm.

    Input:
        nodes: A list of directed graph nodes

    Precondition:
        The input graph is acyclic

    Output:
        A list containing the elements of nodes in an order that puts each node
        before all the nodes it has edges to.
    """
    # Calculate the initial in-degree for each node
    in_degree = {node: len(node.incoming_nodes) for node in nodes}

    # Initialize a deque (double-ended queue) with all nodes that have an in-degree of 0.
    # A deque provides O(1) append and popleft operations, which is ideal for a queue.
    q = collections.deque(node for node in nodes if in_degree[node] == 0)

    # This list will store the final ordered nodes.
    result_ordered_nodes = []

    # Process nodes from the queue
    while q:
        current_node = q.popleft()  # Get the next node from the front of the queue (O(1))
        result_ordered_nodes.append(current_node) # Add to the final result list

        # For each node that 'current_node' points to (its successor)
        for nextnode in current_node.outgoing_nodes:
            # Decrement the in-degree of the successor
            in_degree[nextnode] -= 1

            # If the in-degree of 'nextnode' becomes 0, it means all its predecessors
            # have been processed. Add 'nextnode' to the queue.
            if in_degree[nextnode] == 0:
                q.append(nextnode) # Add to the back of the queue (O(1))

    return result_ordered_nodes

File: adaptive_repair/data/java_programs/TOPOLOGICAL_ORDERING.java
package java_programs;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

// The 'Node' class is assumed to be defined elsewhere and accessible.
// It must have a public method 'getSuccessors()' that returns a List<Node>.
// For Node objects to be used as keys in a HashMap, their equals() and hashCode()
// methods should be properly implemented if object identity is not sufficient for equality.
// For example:
// class Node {
//     // private String id; // Example field
//     // private List<Node> successors; // Example field
//
//     // public Node(String id) { this.id = id; this.successors = new ArrayList<>(); }
//     public List<Node> getSuccessors() {
//         // return this.successors; // Actual implementation
//         return new ArrayList<>(); // Placeholder
//     }
//
//     // @Override
//     // public boolean equals(Object o) { ... }
//     // @Override
//     // public int hashCode() { ... }
// }

public class TOPOLOGICAL_ORDERING {
    public static List<Node> topological_ordering(List<Node> directed_graph) {
        List<Node> orderedNodes = new ArrayList<>();

        // 1. Compute in-degrees for all nodes
        // Initialize all nodes with an in-degree of 0
        Map<Node, Integer> inDegree = new HashMap<>();
        for (Node node : directed_graph) {
            inDegree.put(node, 0);
        }

        // Populate in-degrees by iterating through successors
        for (Node node : directed_graph) {
            for (Node successor : node.getSuccessors()) {
                // Ensure the successor is part of the graph being processed
                // (i.e., it's in the `directed_graph` list). If not, it's an ill-formed graph input.
                if (inDegree.containsKey(successor)) {
                    inDegree.put(successor, inDegree.get(successor) + 1);
                }
            }
        }

        // 2. Initialize the queue (which is `ordered_nodes` in this structure)
        // with nodes that have an in-degree of 0 (no predecessors)
        for (Node node : directed_graph) {
            if (inDegree.get(node) == 0) {
                orderedNodes.add(node);
            }
        }

        int i = 0;
        int listSize = orderedNodes.size(); // Current size of the queue/result list
        
        // 3. Process nodes in the order they are added to ordered_nodes (Kahn's algorithm)
        while (i < listSize) {
            Node node = orderedNodes.get(i); // Get the next node to process (dequeue)
            
            // For each successor of the current node
            for (Node nextNode : node.getSuccessors()) {
                // Decrement the in-degree of the successor
                inDegree.put(nextNode, inDegree.get(nextNode) - 1);
                
                // If the successor's in-degree becomes 0, it means all its predecessors
                // have been processed. Add it to the queue/result list.
                if (inDegree.get(nextNode) == 0) {
                    orderedNodes.add(nextNode);
                    listSize++; // Increment list_size as a new node is added to the queue
                }
            }
            i++;
        }
        
        return orderedNodes;
    }
}